### ISSUE TITLE
<figure> inside <p> is invalid HTML

### ISSUE THREAD

**Issue by cu39** [Evidlo/markdown_captions#6]
According to [HTML Living Standard](https://html.spec.whatwg.org/multipage/grouping-content.html), both `p` and `figure` are _flow content_, and the _content model_ of `p` (permitted elements inside `p`) is only [phrasing content](https://html.spec.whatwg.org/multipage/dom.html#phrasing-content-2). Therefore `p` element having `figure` as its child(ren) is not valid HTML.

There is a [StackOverflow question about this specification](https://stackoverflow.com/questions/5163294/html5-figure-figcaption-inside-a-paragraph-gives-unpredictable-output) in 2011, so `<p><figure><img /></figure></p>` has been invalid during 11 years at least.

IMHO a possible approach is:

1. Find `p` having just an `img` as its child
2. Replace `p` with `figure`
3. Create a `figcaption` element, then append the content of attribute `alt` to it
4. Append `figcaption` to `figure`

Of course it's just a typical case, though.

**Comment by Evidlo** [comment]
This isn't something I have direct control of.  The p tags are are inserted by `python-markdown` itself.

```
>>> md = markdown.Markdown()
>>> md.convert('![alt](img.jpg "title")')
'<p><img alt="alt" src="img.jpg" title="title" /></p>'
```

Maybe consider filing with `python-markdown`.

**Comment by cu39** [comment]
As you said, I currently realize it with my monkey-patching `PostProcessor`, while it's not so difficult to implement it with `TreeProcessor` rather than `InlineProcessor`.

Anyway I assumed the idea behind `markdown_captions` was changing the role of `![]()` syntax from _inline-like_ to _block-like_, because those who choose this extension probably need to place images onto isolated areas (=blocks) dedicated to them (and their captions).

Contrary to my assumption, you said it's not `python-markdown`'s responsibility. I could understand it since `![]()` syntax is originally for inline elements, though validity problem remains.

I have started to think that it would be better, for my use case, to create a new syntax just for `<figure>` blocks. I am going to write it sometime soon.

**Comment by Evidlo** [comment]
I agree that what you say is the idiomatic thing to do, but I'm not sure it's easy to do with a TreeProcessor.

I [used to use TreeProcessor](https://github.com/Evidlo/markdown_captions/commit/afb5b0230920fcf467bf77a5368b7c9cf0279e23), but I had to switch to InlineProcessor in order to be able to capture `{ }` on `![caption](img.jpg title){: .class}`. 

If my memory serves correctly, it was messier to get the caption, file, title, and attr_list fields from the XML tree rather than parsing it myself directly from Markdown in a InlineProcessor.

Maybe some inspiration can be taken from https://github.com/jdittrich/figureAltCaption
